auto_compressor/src/state_saving.rs

```diff
@@ -0,0 +1,260 @@
+pub fn connect_to_database(db_url: &str) -> Result<Client> {
+    let mut builder = SslConnector::builder(SslMethod::tls())?;
+    builder.set_verify(SslVerifyMode::NONE);
+    let connector = MakeTlsConnector::new(builder.build());
+
+    let client = Client::connect(db_url, connector)?;
+    Ok(client)
+}
+
+/// Creates the state_compressor_state and state_compressor progress tables
+///
+/// If these tables already exist then this function does nothing
+///
+/// # Arguments
+///
+/// * `client`        - A postgres client used to send the requests to the database
+pub fn create_tables_if_needed(client: &mut Client) -> Result<()> {
+    let create_state_table = r#"
+        CREATE TABLE IF NOT EXISTS state_compressor_state (
+            room_id TEXT NOT NULL,
+            level_num INT NOT NULL,
+            max_size INT NOT NULL,
+            current_length INT NOT NULL,
+            current_head BIGINT,
+            UNIQUE (room_id, level_num)
+        )"#;
+
+    client.execute(create_state_table, &[])?;
+
+    let create_state_table_indexes = r#"
+        CREATE INDEX IF NOT EXISTS state_compressor_state_index ON state_compressor_state (room_id)"#;
+
+    client.execute(create_state_table_indexes, &[])?;
+
+    let create_progress_table = r#"
+        CREATE TABLE IF NOT EXISTS state_compressor_progress (
+            room_id TEXT NOT NULL,
+            last_compressed BIGINT NOT NULL,
+            UNIQUE (room_id)
+        )"#;
+
+    client.execute(create_progress_table, &[])?;
+
+    Ok(())
+}
+
+/// Retrieve the level info so we can restart the compressor
+///
+/// # Arguments
+///
+/// * `client`        - A postgres client used to send the requests to the database
+/// * `room_id`       - The room who's saved compressor state we want to load
+pub fn read_room_compressor_state(
+    client: &mut Client,
+    room_id: &str,
+) -> Result<Option<(i64, Vec<Level>)>> {
+    // Query to retrieve all levels from state_compressor_state
+    // Ordered by ascending level_number
+    let sql = r#"
+        SELECT level_num, max_size, current_length, current_head, last_compressed
+        FROM state_compressor_state as s
+        LEFT JOIN state_compressor_progress as p ON s.room_id = p.room_id
+        WHERE s.room_id = $1
+        ORDER BY level_num ASC
+    "#;
+
+    // send the query to the database
+    let mut levels = client.query_raw(sql, &[room_id])?;
+
+    // Needed to ensure that the rows are for unique consecutive levels
+    // starting from 1 (i.e of form [1,2,3] not [0,1,2] or [1,1,2,2,3])
+    let mut prev_seen = 0;
+
+    // The vector to store the level info from the database in
+    let mut level_info: Vec<Level> = Vec::new();
+    let mut last_compressed = 0;
+
+    // Loop through all the rows retrieved by that query
+    while let Some(l) = levels.next()? {
+        // read out the fields into variables
+        let level_num: usize = l.get::<_, i32>(0) as usize;
```

Comment:
```suggestion
        let level_num: i32 = l.get("level_num");
```

I'd also be tempted to write it like this, and then cast where necessary later, otherwise the code just looks a bit confusing.

Reply:
I didn't realise you could do the column names! That's much nicer!

I'd rather keep the casting here though, so that `level_num`, `max_size` and `current_length` are never accessible as the wrong type.

Reply:
Have added a comment explaining the confusing type converstions